File: apps/mobile_ios/test/recurrence_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_android/recurrence.dart';

void main() {
  group('expandInstances — weekly recurrence', () {
    test('each weekly instance hour/minute matches the local interpretation of startsAt', () {
      // An event stored in the DB as UTC (e.g. 10:30 UTC).
      // After the fix, each instance carries the *local* hour/minute of that UTC
      // timestamp. Before the fix, it carried the raw UTC hour, which is wrong
      // for any user whose offset != 0.
      final utcStart = DateTime.utc(2026, 4, 1, 10, 30, 0);
      final localStart = utcStart.toLocal();
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.weekly,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 4, 30);

      final instances = expandInstances(e, from, to);
      expect(instances, isNotEmpty,
          reason: 'at least one weekly instance should fall in April');
      for (final inst in instances) {
        // inst is a local (unzoned) DateTime — its .hour/.minute are local fields.
        expect(inst.hour, localStart.hour,
            reason: 'instance hour should be localStart.hour, not the raw UTC hour');
        expect(inst.minute, localStart.minute);
        expect(inst.second, localStart.second);
      }
    });

    test('instances do not precede the original startsAt', () {
      final utcStart = DateTime.utc(2026, 4, 8, 9, 0, 0);
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.weekly,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 4, 30);

      final instances = expandInstances(e, from, to);
      expect(instances, isNotEmpty);
      for (final inst in instances) {
        expect(inst.isBefore(e.startsAt), isFalse,
            reason: 'no instance should precede the event start');
      }
    });

    test('biweekly event produces fewer instances than weekly in the same window', () {
      final utcStart = DateTime.utc(2026, 4, 1, 8, 0, 0);
      final weekly = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.weekly,
      );
      final biweekly = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.biweekly,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 4, 30);

      final wInstances = expandInstances(weekly, from, to);
      final bInstances = expandInstances(biweekly, from, to);
      // Biweekly should produce at most half as many instances.
      expect(bInstances.length, lessThan(wInstances.length));
    });

    test('recurrence_count cap limits the number of instances', () {
      final utcStart = DateTime.utc(2026, 4, 1, 8, 0, 0);
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.weekly,
        count: 2,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 12, 31);

      final instances = expandInstances(e, from, to);
      expect(instances, hasLength(2));
    });

    test('until date stops expansion — no instance falls after until', () {
      final utcStart = DateTime.utc(2026, 4, 1, 8, 0, 0);
      final until = DateTime.utc(2026, 4, 22);
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.weekly,
        until: until,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 12, 31);

      final instances = expandInstances(e, from, to);
      expect(instances, isNotEmpty);
      for (final inst in instances) {
        expect(inst.isAfter(until), isFalse,
            reason: 'no instance should fall after the until date');
      }
    });

    test('non-recurring event returns the single instance if in window', () {
      final utcStart = DateTime.utc(2026, 4, 10, 8, 0, 0);
      final e = EventRecurrence(startsAt: utcStart);

      expect(
        expandInstances(e, DateTime.utc(2026, 4, 1), DateTime.utc(2026, 4, 30)),
        hasLength(1),
      );
      expect(
        expandInstances(e, DateTime.utc(2026, 5, 1), DateTime.utc(2026, 5, 31)),
        isEmpty,
      );
    });
  });

  group('expandInstances — monthly recurrence', () {
    test('monthly event produces one instance per month in window', () {
      final utcStart = DateTime.utc(2026, 4, 5, 10, 0, 0);
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.monthly,
      );

      final from = DateTime.utc(2026, 4, 1);
      final to = DateTime.utc(2026, 6, 30);

      final instances = expandInstances(e, from, to);
      expect(instances, hasLength(3)); // Apr 5, May 5, Jun 5
    });

    test('monthly recurrence_count cap limits instances', () {
      final utcStart = DateTime.utc(2026, 1, 1, 9, 0, 0);
      final e = EventRecurrence(
        startsAt: utcStart,
        freq: RecurrenceFreq.monthly,
        count: 3,
      );

      final from = DateTime.utc(2026, 1, 1);
      final to = DateTime.utc(2026, 12, 31);

      final instances = expandInstances(e, from, to);
      expect(instances, hasLength(3));
    });
  });
}

File: apps/web/src/lib/recurrence.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
	expandInstances,
	nextInstanceAfter,
	describeRecurrence,
} from './recurrence';
import type { Event } from './types';

// Build a partial Event with just enough fields for recurrence math.
// Casts the rest to `unknown as Event` — the recurrence helpers only
// touch `starts_at`, `recurrence_freq`, `recurrence_byday`,
// `recurrence_until`, and `recurrence_count`.
function ev(partial: Partial<Event>): Event {
	return partial as unknown as Event;
}

test('expandInstances — non-recurring event returns its single instance when in window', () => {
	const e = ev({ starts_at: '2026-04-10T08:00:00Z' });
	const inWindow = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-04-30T23:59:59Z'),
	);
	assert.equal(inWindow.length, 1);

	const outOfWindow = expandInstances(
		e,
		new Date('2026-05-01T00:00:00Z'),
		new Date('2026-05-31T23:59:59Z'),
	);
	assert.equal(outOfWindow.length, 0);
});

test('expandInstances — weekly event produces multiple instances in a month', () => {
	const e = ev({
		starts_at: '2026-04-01T08:00:00Z',
		recurrence_freq: 'weekly',
	});
	const out = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-04-30T23:59:59Z'),
	);
	// Apr 1 (Wed), Apr 8, Apr 15, Apr 22, Apr 29 → 5 instances. Allow
	// 4-5 to absorb timezone-edge slip on the loader's host.
	assert.ok(out.length >= 4 && out.length <= 5, `got ${out.length} instances`);
});

test('expandInstances — biweekly produces fewer instances than weekly in the same window', () => {
	const start = '2026-04-01T08:00:00Z';
	const weekly = ev({ starts_at: start, recurrence_freq: 'weekly' });
	const biweekly = ev({ starts_at: start, recurrence_freq: 'biweekly' });
	const from = new Date('2026-04-01T00:00:00Z');
	const to = new Date('2026-04-30T23:59:59Z');
	const w = expandInstances(weekly, from, to);
	const b = expandInstances(biweekly, from, to);
	assert.ok(b.length < w.length, `biweekly ${b.length} should be fewer than weekly ${w.length}`);
});

test('expandInstances — recurrence_count caps the number of instances', () => {
	const e = ev({
		starts_at: '2026-04-01T08:00:00Z',
		recurrence_freq: 'weekly',
		recurrence_count: 3,
	});
	const out = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-12-31T23:59:59Z'),
	);
	assert.equal(out.length, 3);
});

test('expandInstances — recurrence_until stops the expansion', () => {
	const until = '2026-04-22T08:00:00Z';
	const e = ev({
		starts_at: '2026-04-01T08:00:00Z',
		recurrence_freq: 'weekly',
		recurrence_until: until,
	});
	const out = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-12-31T23:59:59Z'),
	);
	assert.ok(out.length > 0);
	const untilMs = new Date(until).getTime();
	for (const d of out) {
		assert.ok(d.getTime() <= untilMs, `instance ${d.toISOString()} after until`);
	}
});

test('expandInstances — monthly produces one instance per month in window', () => {
	const e = ev({
		starts_at: '2026-04-05T10:00:00Z',
		recurrence_freq: 'monthly',
	});
	const out = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-06-30T23:59:59Z'),
	);
	assert.equal(out.length, 3);
});

test('expandInstances — instances do not precede the original starts_at', () => {
	const startsAt = '2026-04-08T09:00:00Z';
	const e = ev({ starts_at: startsAt, recurrence_freq: 'weekly' });
	const out = expandInstances(
		e,
		new Date('2026-04-01T00:00:00Z'),
		new Date('2026-04-30T23:59:59Z'),
	);
	const startMs = new Date(startsAt).getTime();
	for (const d of out) {
		assert.ok(d.getTime() >= startMs, `instance ${d.toISOString()} precedes starts_at`);
	}
});

test('expandInstances — monthly with recurrence_count caps instances', () => {
	const e = ev({
		starts_at: '2026-01-01T09:00:00Z',
		recurrence_freq: 'monthly',
		recurrence_count: 4,
	});
	const out = expandInstances(
		e,
		new Date('2026-01-01T00:00:00Z'),
		new Date('2026-12-31T23:59:59Z'),
	);
	assert.equal(out.length, 4);
});

test('nextInstanceAfter — returns the next instance after a cutoff', () => {
	const e = ev({
		starts_at: '2026-04-01T08:00:00Z',
		recurrence_freq: 'weekly',
	});
	const next = nextInstanceAfter(e, new Date('2026-04-10T00:00:00Z'));
	assert.ok(next != null);
	// The first weekly instance after Apr 10 should be Apr 15 (Wed).
	assert.equal(next!.toISOString().slice(0, 10), '2026-04-15');
});

test('nextInstanceAfter — returns null past recurrence_until', () => {
	const e = ev({
		starts_at: '2026-04-01T08:00:00Z',
		recurrence_freq: 'weekly',
		recurrence_until: '2026-04-30T23:59:59Z',
	});
	const next = nextInstanceAfter(e, new Date('2026-05-01T00:00:00Z'));
	assert.equal(next, null);
});

test('describeRecurrence — null freq is "One-off event"', () => {
	assert.equal(describeRecurrence(null, []), 'One-off event');
	assert.equal(describeRecurrence(null, null), 'One-off event');
});

test('describeRecurrence — monthly is "Repeats monthly"', () => {
	assert.equal(describeRecurrence('monthly', null), 'Repeats monthly');
	assert.equal(describeRecurrence('monthly', ['MO']), 'Repeats monthly');
});

test('describeRecurrence — weekly with byday lists days in ISO order', () => {
	// Even when input order is shuffled, output order is MO,TU,...,SU.
	assert.equal(describeRecurrence('weekly', ['WE', 'MO']), 'Every week · Mon, Wed');
});

test('describeRecurrence — biweekly without byday is bare', () => {
	assert.equal(describeRecurrence('biweekly', null), 'Every other week');
	assert.equal(describeRecurrence('biweekly', []), 'Every other week');
});
